src/lib.rs

```diff
@@ -199,19 +199,36 @@ mod zerocopy {
 ///
 /// # Safety
 ///
-/// If `T: FromZeroes`, then unsafe code may assume that it is sound to treat
-/// any initialized sequence of zero bytes of length `size_of::<T>()` as a `T`.
+/// *This section describes what is required in order for `T: FromZeroes`, and
+/// what unsafe code may assume of such types. `#[derive(FromZeroes)]` only
+/// permits types which satisfy these requirements. If you don't plan on
+/// implementing `FromZeroes` manually, and you don't plan on writing unsafe
+/// code that operates on `FromZeroes` types, then you don't need to read this
+/// section.*
+///
+/// If `T: FromZeroes`, then unsafe code may assume that:
+/// - ...it is sound to treat any initialized sequence of zero bytes of length
+///   `size_of::<T>()` as a `T`.
+/// - ...given `b: &[u8]` where `b.len() == size_of::<T>()` and `b` contains
+///   only zero bytes, it is sound to construct a `t: &T` at the same address as
+///   `b`, and it is sound for both `b` and `t` to be live at the same time.
+///
 /// If a type is marked as `FromZeroes` which violates this contract, it may
 /// cause undefined behavior.
 ///
-/// If a type has the following properties, then it is safe to implement
+/// If a type has the following properties, then it is sound to implement
 /// `FromZeroes` for that type:
-/// - If the type is a struct, all of its fields must implement `FromZeroes`
+/// - If the type is a struct, all of its fields must satisfy the requirements
+///   to be `FromZeroes` (they do not actually have to be `FromZeroes`).
 /// - If the type is an enum, it must be C-like (meaning that all variants have
 ///   no fields) and it must have a variant with a discriminant of `0` (see [the
-///   reference] for a description of how discriminant values are chosen)
+///   reference] for a description of how discriminant values are chosen).
+/// - The type must not contain any [`UnsafeCell`]s (this is required in order
+///   for it to be sound to construct a `&[u8]` and a `&T` to the same region of
+///   memory).
```

Comment:
I wasn't sure how to phrase this since we don't implement `FromZeroes` for `Option<&...>` today. How does this sound?

> The type must not contain any `UnsafeCell`s (this is required in order for it to be sound to construct a `&[u8]` and a `&T` to the same region of memory). The type may contain references or pointers to `UnsafeCell`s so long as those values can themselves be initialized from zeroes (`FromZeroes` is not currently implemented for, e.g., `Option<&UnsafeCell<_>>`, but it could be one day).

Reply:
That sounds good, other than the generally applicable technical writing advice to avoid parentheticals longer than a word or two.

Reply:
Gonna merge this for now; feel free to follow up with more discussion here and I'd be happy to update the docs in a follow-up PR.